src/ua_securechannel.c

```diff
@@ -230,189 +248,160 @@ void UA_SecureChannel_revolveTokens(UA_SecureChannel* channel) {
-
-    // TODO: Potentially introduce error handling to the encryption methods???
-
-    // Bring back hidden bytes for signature
-    if(channel->securityMode == UA_MESSAGESECURITYMODE_SIGN ||
-       channel->securityMode == UA_MESSAGESECURITYMODE_SIGNANDENCRYPT) {
-        dst->length += securityPolicy->symmetricModule.signingModule.signatureSize;
+    /* An error occurred. Overwrite the current chunk with an abort message. */
+    if(ci->errorCode != UA_STATUSCODE_GOOD) {
+        *buf_pos = buf_body_start;
+        UA_String errorMsg = UA_STRING_NULL;
+        UA_encodeBinary(&ci->errorCode, &UA_TYPES[UA_TYPES_UINT32],
+                        buf_pos, buf_end, NULL, NULL);
+        UA_encodeBinary(&errorMsg, &UA_TYPES[UA_TYPES_STRING],
+                        buf_pos, buf_end, NULL, NULL);
+        ci->final = true; /* Mark chunk as finished */
     }
 
-    // Pad the message
+    /* Pad the message. The bytes for the padding and signature were removed
+     * from buf_end before encoding the payload. So we don't check here. */
+    // TODO: What about just SIGN?
     if(channel->securityMode == UA_MESSAGESECURITYMODE_SIGNANDENCRYPT) {
-        // Bring back hidden bytes for padding and extra padding
-        dst->length += 2;
-
+        size_t payloadLength = (uintptr_t)buf_body_end - (uintptr_t)ci->messageBuffer.data;
         UA_Byte paddingSize = 0;
         UA_Byte extraPaddingSize = 0;
-        securityPolicy->symmetricModule.calculatePadding(securityPolicy, bytesToWrite, &paddingSize, &extraPaddingSize);
+        securityPolicy->symmetricModule.calculatePadding(securityPolicy, payloadLength,
+                                                         &paddingSize, &extraPaddingSize);
         UA_UInt16 totalPaddingSize = (UA_UInt16)((extraPaddingSize << 8) | paddingSize);
 
-        for(UA_UInt16 i = 0; i <= totalPaddingSize; ++i) // This is <= because the paddingSize byte also has to be written.
-        {
-            dst->data[offset] = paddingSize;
-            ++offset;
+        /* This is <= because the paddingSize byte also has to be written. */
+        for(UA_UInt16 i = 0; i <= totalPaddingSize; ++i) {
+            **buf_pos = paddingSize;
+            (*buf_pos)++;
         }
 
         if(extraPaddingSize > 0) {
-            dst->data[offset] = extraPaddingSize;
-            ++offset;
+            **buf_pos = extraPaddingSize;
+            (*buf_pos)++;
         }
     }
 
-    if(ci->messageSizeSoFar + offset > connection->remoteConf.maxMessageSize &&
-       connection->remoteConf.maxMessageSize > 0)
-        ci->errorCode = UA_STATUSCODE_BADRESPONSETOOLARGE;
-    if(++ci->chunksSoFar > connection->remoteConf.maxChunkCount &&
-       connection->remoteConf.maxChunkCount > 0)
-        ci->errorCode = UA_STATUSCODE_BADRESPONSETOOLARGE;
+    /* The total message length */
+    UA_UInt32 pre_sig_length = (UA_UInt32)((uintptr_t)buf_body_end - (uintptr_t)ci->messageBuffer.data);
```

Comment:
We want to sign everything from the beginning to the end including the padding, so shouldn't this be buf_pos?